homeassistant/components/bluetooth/__init__.py

```diff
@@ -0,0 +1,267 @@
+"""The bluetooth integration."""
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+import dataclasses
+from enum import Enum
+import fnmatch
+import logging
+from typing import Final
+
+from bleak import BleakError
+from bleak.backends.device import MANUFACTURERS, BLEDevice
+from bleak.backends.scanner import AdvertisementData
+
+from homeassistant import config_entries
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import (
+    CALLBACK_TYPE,
+    Event,
+    HomeAssistant,
+    callback as hass_callback,
+)
+from homeassistant.data_entry_flow import BaseServiceInfo
+from homeassistant.helpers import discovery_flow
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.loader import BluetoothMatcher, async_get_bluetooth
+
+from . import models
+from .const import DOMAIN
+from .models import HaBleakScanner
+from .usage import install_multiple_bleak_catcher
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BluetoothScanningMode(Enum):
+    """The mode of scanning for bluetooth devices."""
+
+    PASSIVE = "passive"
+    ACTIVE = "active"
+
+
+SCANNING_MODE_TO_BLEAK = {
+    BluetoothScanningMode.ACTIVE: "active",
+    BluetoothScanningMode.PASSIVE: "passive",
+}
+
+LOCAL_NAME: Final = "local_name"
+SERVICE_UUID: Final = "service_uuid"
+MANUFACTURER_ID: Final = "manufacturer_id"
+MANUFACTURER_DATA_FIRST_BYTE: Final = "manufacturer_data_first_byte"
+
+
+@dataclasses.dataclass
+class BluetoothServiceInfo(BaseServiceInfo):
+    """Prepared info from bluetooth entries."""
+
+    name: str
+    address: str
+    rssi: int
+    manufacturer_data: dict[int, bytes]
+    service_data: dict[str, bytes]
+    service_uuids: list[str]
+
+    @classmethod
+    def from_advertisement(
+        cls, device: BLEDevice, advertisement_data: AdvertisementData
+    ) -> BluetoothServiceInfo:
+        """Create a BluetoothServiceInfo from an advertisement."""
+        return cls(
+            name=advertisement_data.local_name or device.name or device.address,
+            address=device.address,
+            rssi=device.rssi,
+            manufacturer_data=advertisement_data.manufacturer_data,
+            service_data=advertisement_data.service_data,
+            service_uuids=advertisement_data.service_uuids,
+        )
+
+    @property
```

Comment:
```suggestion
    @cached_property
```

```
from functools import cached_property
```

https://docs.python.org/3/library/functools.html#functools.cached_property